### DISCUSSION TITLE
Create a usecase diagram for a parking lot management system

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#24] [General]
What?
Identify the actors and usecases to generate a usecase UML diagram for a parking lot.

Why?
1. Helps understand and document the requirements.
2. Helps identifying the APIs to create.

How?
Fill out the worksheet [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/04-design-a-parking-lot-worksheet.md#use-case-diagrams).

**Comment by VishalGupta25** [comment]
@startuml
left to right direction
actor "Entry Parking Attendant" as NPA 
actor "Exit Parking Attendant" as XPA 
actor "Vehicle Driver" as VD

rectangle Parkilot {
  usecase "Issue Ticket" as IT
  usecase "Check Available Slots" as CAS

  usecase "Register the Slot " as RS
  usecase "Park Car At Slot" as PCS 
  usecase "Checkout Vehicle" as CV
  usecase "Check Total Parking Time" as COPT
  usecase "Generate Parking Bill" as GPB
  usecase "Make Payment" as MP
  usecase "Free the Slot " as FS



  
}
NPA --> IT
NPA --> CAS
NPA --> RS
VD--> PCS
VD --> CV 
XPA --> COPT
XPA --> GPB
VD --> MP
XPA --> FS
@enduml

**Reply by VishalGupta25** [reply]
![image](https://github.com/kanmaytacker/design-questions/assets/40258077/671e8f56-8ae2-4d01-bee2-555d5f9ff630)

**Reply by kanmaytacker** [reply]
This is great @VishalGupta25. A few things to add:
1. You would need a user/customer actor since they can pay directly online
2. You can also have an admin actor for adding the entities
3. Think of includes and extends keywords

**Reply by VishalGupta25** [reply]
@startuml
actor "Entry Parking Attendant" as NPA 
actor "Exit Parking Attendant" as XPA 
actor "Vehicle Driver" as VD

rectangle Parkinglot {

left to right direction
  usecase "Issue Ticket" as IT
  usecase "Check Available Slots" as CAS
  usecase "Register the Slot " as RS
  usecase "Park Car At Slot" as PCS 
  usecase "Checkout Vehicle" as CV 
  usecase "Check Total Parking Time" as COPT
  usecase "Generate Parking Bill" as GPB
  usecase "Make Payment" as MP
  usecase "Free the Slot " as FS 
  IT --> CAS: include
  CAS -> RS: include 
 
}
NPA --> IT : request
VD--> PCS
VD --> CV 
XPA --> COPT
XPA --> GPB
VD --> MP
XPA --> FS
@enduml

**Comment by animesh799** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/93705578/4a98b8c9-29be-4f60-9f52-3a25175827bf)

**Reply by kanmaytacker** [reply]
@animesh799 The use-cases should be the actual functionalities. Payment or rather pay is a functionality but park will not be. You might also need usecases for generating the bill.

**Comment by gosharmatelia** [comment]
actor "Visitor" as vs 
actor "Operator" as op

ParkingLot parkingLot{
   usecase "checkAvailability" as UC1
   usecase "allot slot"       as UC2
   usecase "fill the slot and update in display board" as UC3
   usecase "select Payment mode" UC4 
   usecase "calculate payment and provide bill" UC5
   usecase "update the slot as empty and udpate the display board" UC6
}

op --> UC1
op --> UC2
op --> UC3
op --> UC4
op --> UC5
op --> UC6

**Comment by levinrockx** [comment]
![parking](https://github.com/kanmaytacker/design-questions/assets/10922924/c7093f76-f3e9-45c9-8f71-60059c9a7ff5)

**Comment by venu-ta-philips** [comment]
![parking lot use case](https://github.com/kanmaytacker/design-questions/assets/125430291/8c729c82-b1ec-475c-96a1-dcee5de7f1f9)

**Comment by vishwaraths** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/11187838/acc6155f-3a8a-480e-8b67-650044d38e78)

**Comment by veersingh** [comment]
Actor:
Driver
System Admin

Use cases:
1. Entry
2. Booking Parking
3.  Payment
4.  Ticket generate
5.   Display update
6.  Exit

**Comment by MauryaBrij** [comment]
Actors:
	- Vehicle
	- parkingLot
	- Admin/parkingAttendent
	- Ticket
	- Spot
	- floor
	- paymentCounter
	- entry & exit gates
UseCases:
	- checkSpotAvailability
	- displaySpots
	- createTicket
	- feeCalculation
	- feePayment

**Comment by deepakkumarpallai** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/143584040/cf493602-ddb3-4351-9b3f-cb0b852047b4)

**Comment by krishnak15** [comment]
![WhatsApp Image 2023-10-11 at 22 02 44 (1)](https://github.com/kanmaytacker/design-questions/assets/113184344/8a3ef387-0610-41fc-a210-2b680b4e25e3)

**Comment by MakaBaca** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/42783016/3388be7f-8af7-4316-9e8c-b49cba3ccf68)

**Comment by Abhilash-du** [comment]
Actors:-
	1, Customer (anyone who parks vehicle)
	2. Admin (Manages parking lot system, add/remove parking area, add floors)
	3. Ticket Collector (collect ticket and deal with customer for payment)


Use cases:
Actor 1
Name of the actor -  Customer
1. Customer enters the parking lot by taking a ticket
2. Customer exits parking lot by paying ticket
3. Customer Paying the charge for parking based on vehicle and timing


Actor 2
Name of the actor -  Admin
1. Adding parking, removing parking spaces based on occupancy
2. Can Set Parking rates (if required)
3. Can add new floors (if required)

Actor 3
Name of the actor -  Ticket Collector
1. Issue ticket to customer
2. Collect payment from customer

**Comment by Thejas997** [comment]
# User:
- Request for parking slot
- Request for bill
- Pay for parking
  - Exit Counter
      - Cash
  - Payment Counter
      - Credit Card
      - UPI
  - Online
      - Credit Card
      - UPI

# Parking Attendant:
- Check slot availability
- Issue ticket
- Issue bill
- Accept payment

**Comment by arafatoffcl** [comment]
![Parking_slot](https://github.com/kanmaytacker/design-questions/assets/147293525/28352448-4043-4eea-aa11-0a2d4c5694f5)

**Reply by VishalGupta25** [reply]
HI Can you share your uml code here?

**Comment by aniketankalikar** [comment]
Actors
-----------

Floor
Points (Exit and Entry) / Attendant
User
Car [Large / Small/ Medium]
Parking Spot
Display Status



Use Cases

1. User Enters Entry Point
2. User collects Ticket
3. Check for Display Baord for available Slot
4. User Enters Floor with Available spot
4. Find a spot for the car
5. Move towards Exit Point depending on if you are paying CASH or Dedicated Booth
6. Attendant calculates the FEES based on the type of CAR
7. Choose the Type of payment
8. Pay Fees
9. Exit from the exit point

**Comment by shraddha1194** [comment]
Actor1
Customer:
1. Enter Parking
2. Park
3. Collect ticket
4. Make payment
5. exit parking
6. check slot

Actor2:
Operator:
1. Check slots
2. Update slot status
3. Provide ticket
4. collect payment

**Comment by nage1234** [comment]
![Uploading 1697042275244.jpg…]()

**Comment by nage1234** [comment]
![1697042275244](https://github.com/kanmaytacker/design-questions/assets/88926279/3ecad50f-876e-4e18-aed6-36627db9c054)

**Comment by PrinceKushwaha1** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/136918412/9d4bce63-8362-4142-97b3-879d1248efbd)

**Comment by shaileshexp** [comment]
![parking_lot](https://github.com/kanmaytacker/design-questions/assets/33438762/45b6261a-e26a-4802-8dce-b687e2f35701)

**Comment by pramodpatil812** [comment]
@startuml
'https://plantuml.com/use-case-diagram
left to right direction

actor User as U
actor Admin as A
actor Operator as O
actor Customer as C

A -> U : extends
O -> U : extends
C -> U : extends

rectangle "Parking Lot Management" {
  usecase "Create Operator User" as UC0
  usecase "Login" as UC1
  usecase "Enter Parking Entry" as UC2
  usecase "Check Parking Availability" as UC3
  usecase "Allot vehicle parking" as UC4
  usecase "Issue Ticket" as UC5
  usecase "Park at the alloted spot" as UC6
  usecase "Calculate payable amount" as UC8
  usecase "Issue bill" as UC9
  usecase "Pay" as UC10
  usecase "UPI" as UC11
  usecase "CASH" as UC12
  usecase "Check the Slot" as UC13
  usecase "Free Slot" as UC14
  usecase "CRUD on Slots" as UC15
  usecase "CRUD on  Slot Fee" as UC16
  UC11-->UC10 : extends
  UC12->UC10 : extends
  UC5->UC13 : include
}

A --> UC0
A --> UC15
A --> UC16
O --> UC1
C --> UC2
O --> UC3
O --> UC4
O --> UC5
C --> UC6
O --> UC8
O --> UC9
C --> UC10
O --> UC14
@enduml
![parkinglot](https://github.com/kanmaytacker/design-questions/assets/5285389/00fb6313-56d1-432c-a4b3-230512d58f55)

**Comment by agarwalprerna643** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/51155049/061a0b7c-0b3f-4384-b6ec-dc89c871b87a)

**Comment by anandayush81** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/39666801/0e0c62a1-a29b-4690-bcbc-67e2110656f5)

**Comment by ris1b** [comment]
![img_20231011_221048_720](https://github.com/kanmaytacker/design-questions/assets/101881146/8b9a3dcb-e295-43bb-9431-9755e2dc532c)

**Comment by Harikrishnan19** [comment]
> What? Identify the actors and usecases to generate a usecase UML diagram for a parking lot.
> 
> Why?
> 
> 1. Helps understand and document the requirements.
> 2. Helps identifying the APIs to create.
> 
> How? Fill out the worksheet [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/04-design-a-parking-lot-worksheet.md#use-case-diagrams).

@startuml
left to right direction
actor "driver" as d
actor "inspector" as i
actor "admin" as a
rectangle "Parking Lot" {
  usecase "Get ticket" as UC1
  usecase "Park at the given lot" as UC2
  usecase "Get charges" as UC3
  usecase "calculate charges" as UC9
  usecase "Pay" as UC4
  usecase "Pay by cash" as UC6
  usecase "Pay by card" as UC7
  usecase "Pay by upi" as UC8
  usecase "Display parking lots" as UC5
}
d --> UC1
d --> UC2
d --> UC4
i --> UC3
a --> UC5
i --> UC5
UC3 .> UC9
UC4 --|> UC6
UC4 --|> UC7
UC4 --|> UC8

@enduml
![parking-lot-use-case-diagram](https://github.com/kanmaytacker/design-questions/assets/67498840/28adc130-6013-431d-a9b9-ac4253ee1b17)

**Comment by annant05** [comment]
![Screenshot 2023-10-11 221224](https://github.com/kanmaytacker/design-questions/assets/21981463/2798d00e-2145-4d8f-84ce-722545698600)

**Comment by sandy2608** [comment]
@startuml
left to right direction
actor "Vehicle" as fc
rectangle "Parking Lot" {
  usecase "Slot Available" as UC1
  usecase "Get Ticket" as UC2
  usecase "Pay Bill" as UC3
}
fc --> UC1
fc --> UC2
fc --> UC3
@enduml

**Comment by hey-vishal** [comment]
@startuml
left to right direction
actor "User" as u
left to right direction
actor "Operator" as op
rectangle Parking_Lot {
  usecase "Check Slot" as UC1
  usecase "Generate ticket" as UC2
  usecase "Get ticket" as UC3
  usecase "Park vehicle" as UC4
  usecase "Check Ticket" as UC5
  usecase "Make Payment" as UC6
  usecase "Verify Payment" as UC7
  usecase "Exit" as UC8
}
u --> UC3
u --> UC4
u --> UC6
u --> UC8
op --> UC1
op --> UC2
op --> UC5
op --> UC7
@enduml

**Comment by rshankarj** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/13399867/5eec54df-7063-4186-9679-c850073ac18c)

**Comment by sanketh2222** [comment]
![pl](https://github.com/kanmaytacker/design-questions/assets/41993685/191a76c9-6f6e-42f9-abe8-af6c7a9dcc46)

**Comment by sivarenjinitm** [comment]
![XP71JW8n48RlVOevmW4Vm8C1YJSJIJJSQtUMdUniXtJ_5HlZkri9CI0r7llFbs_QlyiHFc4Qr2ct81WbsNMWHX87Y4Nd0opHbZi9oa-7oEaKRNpg9UuU3VU6gsm5S6nyXFiM-RXJFfi-vn4yq8Sxzz9YGRE72FxnK47d5UPTS8bStEN8h3iEFQtUlAX_5H (1)](https://github.com/kanmaytacker/design-questions/assets/104749750/15de612c-5f78-4f1f-93d4-ead4cfb37b4a)

**Comment by anisdeepam** [comment]
[Parking Lot.drawio.pdf](https://github.com/kanmaytacker/design-questions/files/12872748/Parking.Lot.drawio.pdf)

**Comment by soorya-1992** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/142311696/db6230b6-e75a-4503-a9ea-efc967a79ebc)

**Comment by artipra** [comment]
![parking-lot-usecase-diagram](https://github.com/kanmaytacker/design-questions/assets/12079701/cc7abf39-984c-4669-9b78-c299c0744f14)
parking lot use case diagram
